### ISSUE TITLE
Support for typescript generation

### ISSUE THREAD

**Issue by dar2355** [kaitai-io/kaitai_struct#542]
Currently, the only technologies I have knowledge of how to use on the web is javascript and typescript.

The only one that kaitai struct is able to compile into at the moment is javascript. While it's still significantly better than nothing, utilizing the generated javascript is difficult (or at least very tedious) compared to other programming languages.

I attribute this tedium to the lack of autocompletion that javascript offers. IDEs such as VSCode  are unable to provide "smart" autocompletion for javascript. However, it is able to do so for typescript. This removes the necessity to memorize every little part of the structure when writing code to work with the structures.

Here is an example of what I'm talking about with autocompletion: ![image](https://user-images.githubusercontent.com/3662774/54934400-d429b280-4ef4-11e9-96a9-12fe5bb632f3.png).

the suggestion "mainHeader" would not show up if this were created in plain javascript. However, because it's a snipped of something custom I developed in typescript, it shows up. This would significantly cut-down on the development time of web-based projects utilizing Kaitai Struct's generated code.

--- 

**TL;DR:** 
- **GOAL**: 
to generate typescript code (or at least definitions)

- **HOW THIS WOULD HELP**:
It would cut down on development time of web applications utilizing Kaitai Struct.

**Comment by GreyCat** [comment]
Cc @jchv

**Comment by GreyCat** [comment]
Take a look at http://doc.kaitai.io/new_language.html, if you haven't seen it yet — it gives the overall idea of scope.

That said, probably TS target should be easy enough, being mostly either a copy-pasted clone of JavaScript target, or even an extension of thereof (as with `cpp_stl_98` vs `cpp_stl_11`), so either way it should be relatively quick to solve.

If we could reuse existing tests for JS to some extent, that would be a great start. If you want to contribute, I'd suggest to start with creation of that test infrastructure first — i.e. at bare minimum we'd want:

* `ci-typescript` in [tests repo](https://github.com/kaitai-io/kaitai_struct_tests), which will run all the tests and produce a JSON result file (akin to [ci-javascript](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/ci-javascript)).
* `prepare-typescript` in [ci_targets repo](https://github.com/kaitai-io/ci_targets) to install all the prerequisites and prepare the environment prior to running these tests (akin to [prepare-javascript](https://github.com/kaitai-io/ci_targets/blob/master/prepare-javascript)).
* It we'll need dedicated set of tests, we'll have to clone [spec/javascript in tests repo](https://github.com/kaitai-io/kaitai_struct_tests/tree/master/spec/javascript) and convert it to `spec/typescript`.

Once that will be done, we'll start a typescript story in CI and it will get its own column in https://ci.kaitai.io/

**Comment by jchv** [comment]
I've been looking into it already, though I am just learning Scala now, so it may be a while before I am able to get code good enough to contribute. :)

So, here's the deal. Today, Kaitai Struct generates a UMD module containing old-school prototype-based classes. We could go two directions with this:

1. We could generate .d.ts files to accompany the UMD module. This would give you all of the editor goodies you want without having to re-haul the compiler; you could implement .d.ts as its own language or something. Seems like too much of a hack for my tastes, though.

2. We could generate TypeScript directly, of course. This is the one I'm most interested in working on. It has the added benefit that mistakes in the compiler would show up immediately and could be tested with automated testing.

Right now I think the best approach here is to make TypeScript just be a special case of the JavaScript compiler. To get there, we need to modernize the output so that it can be typed. So, we need to first make the compiler support outputing ES modules and ES classes. It really wouldn't hurt to add support for other module systems as well, like CommonJS.

Adding new module systems should not be too hard. It's basically a matter of using different templates based on which one we're in. For UMD, you want a `return Thing`, for ES you want `export default Thing;` and for CommonJS you want `export.modules = Thing;` or so. (Whether or not to use a default export is questionable, but it is more or less the current behavior. For ES modules, it might make more sense to just export differently.) This issue is already being tracked, sort of, at #267.

ES classes are a little bit different, but it's still roughly the same idea: mostly the same code generation, with different templates. I created an issue for this at #539.

The final thing would be to add type annotations everywhere. I don't think that'll be too difficult to do, either. There might be some weird things to handle, like enums. But most things, I think, will not need to be re-hauled.

One other thing we need in order to be able to make a TypeScript target is a TypeScript runtime component. I've created a PR that adds typings (https://github.com/kaitai-io/kaitai_struct_javascript_runtime/pull/10) which would be enough to get started, though it would be ideal in the end if it was just converted to TypeScript (A process that will be quite easy, but I'd like to see more testing in place before it's done, to ensure that such a swap does not cause any compatibility issues.) I think it's easy enough to see the benefit of switching it to TS, since the TypeScript compiler was already able to catch a minor bug in the code, but it's very important that it doesn't break the API or browser compatibility (within reason.)

**Comment by dar2355** [comment]
I agree - I would like to generate the typescript directly. I've just had trouble wrapping my head around how to get started (hence why I tried for the past 6 months to try and recreate kaitai struct my own way to no avail lol).

Even more, I'd love to add support for re-serializing the structures back into binary once modifications have been made, but that is a whole other giant thing completely. For now, yeah - I'd love to get javascript / typescript working in a modern way to get those awesome IDE goodies. 

Alternatively, we could _in-theory_ use flow, but that, like the .d.ts files feels more like a hack than an actual solution.

**Comment by jchv** [comment]
@dar2355 

>Even more, I'd love to add support for re-serializing the structures back into binary once modifications have been made, but that is a whole other giant thing completely. For now, yeah - I'd love to get javascript / typescript working in a modern way to get those awesome IDE goodies.

You know, it never occurred to me that Kaitai doesn't support this. It's unfortunately a huge bag of worms, though; the ksy format would probably need to go through a major iteration to handle all of the tricky cases for the encode as it does for the decode :( It would probably be hard to get compex formats like TTF working both directions.

This is totally a different bug, but a bug that should be created I think.

>Alternatively, we could in-theory use flow, but that, like the .d.ts files feels more like a hack than an actual solution.

I think that even if you wanted to support Flow it'd be better to go this way, personally; ES classes and modules are typed in a manner that is pretty similar in both TypeScript and FlowType.

**Comment by FireyFly** [comment]
As a JS and occasionally Flow user, I'd like to chime in a bit. In both these languages, type definition files is how you'd usually specify types for external dependencies.

To me, since AFAIK the parser itself isn't meant to be edited by hand after generation (but rather to be imported and interfaced with, and regenerated as the grammar changes), it makes sense to take a similar approach with definition files for TypeScript and Flow. It also means there's only one core code path to maintain (generating ES5 code, so it's compatible with both new and old JS), and depending on flags provided, a TypeScript or Flow typing file could be generated adjacent to it.

I think whilst it would certainly be possible to offer ES5, modern ES, modern ES with TypeScript typing, and modern ES with Flow typing backends, to me at least that feels a bit less clean.

**Comment by GreyCat** [comment]
@jchv 
> This is totally a different bug, but a bug that should be created I think.

#27?

**Comment by jchv** [comment]
@GreyCat 
> #27 

Thanks, watching that now.

@FireyFly 
>I think whilst it would certainly be possible to offer ES5, modern ES, modern ES with TypeScript typing, and modern ES with Flow typing backends, to me at least that feels a bit less clean.

I'm looking for end-to-end type safety. I feel for my use cases, having everything in ES modules and classes is easier because my bundler, be it Webpack or Parcel or what have you, can then handle the transpilation and module wiring, and there's no need to worry about the type definitions lagging.

If you generate `.d.ts` files, there's nothing preventing the generated types from being incorrect, or the generated code from misusing the types. But, generating type safe code is clearly possible since there are plenty of other statically typed languages that Kaitai can generate code for just fine. So I think the aim would be to generate strict code that is able to pass the type checkers.

FWIW, I'm not suggesting that the current output should ever be removed, just that offering more modern output would be desirable to me as an optional feature, similar to offering variations depending on C++ version. I could be wrong but I don't believe the gap between the output now and the ideal TypeScript output is actually huge, and I can sort-of prove that by doing the work manually.

I haven't used Flow in a bit, but last I checked FlowType and TypeScript are not very different syntactically. On the other hand, their _type definition_ format is quite different, which may actually make out of band type definitions harder if you wanted to share code.

**Comment by FireyFly** [comment]
Fair enough, that seems reasonable I suppose, yeah.

**Comment by dar2355** [comment]
ok so I created a file format (super smash bros brawl's RSTM (music) files) and have converted it to the equivalent typescript code. This has many "implicit any"s but this should display an idea of what the target should / could look like

[here's the gist](https://gist.github.com/dar2355/e56691b621dee6534f9e0b9365b23fd5)

**Comment by GreyCat** [comment]
@dar2355 Could you actually elaborate what will it take to get us TS testing in the CI? Is it just "we need nodejs + some dependencies + some launcher script"? If so, you chance you can start these?

**Comment by dar2355** [comment]
I'll need to look closer into the proper ways of adding languages, but for the most part it seems we can just extend and modify the javascript. I'll need to look into how you go about doing the different versions of C++

I have a super busy academic week (finals are soon), but I should be able to look into it much more within the next week or two

**Comment by Koulwa** [comment]
Any luck with this? TS with a more complicated packet would be a real huge productivity booster.

**Comment by Theaninova** [comment]
In case anyone is interested, I've put together some rudimentary type generation for a vite plugin that allows direct import of ksy files.

https://github.com/Theaninova/vite-plugin-kaitai/blob/master/src/kaitai-type-generator.ts

**Comment by nullableVoidPtr** [comment]
Thanks @Theaninova for the PR! I've an incomplete ground-up implementation from the original work from fudgepop at [this compiler fork](https://github.com/nullableVoidPtr/kaitai_struct_compiler/tree/TypeScriptIntegration) and [runtime](https://github.com/nullableVoidPtr/kaitai_struct_typescript_runtime). I've attempted to shoe in BigInt, but it seems there would need to be a lot of Kaitai Ast transforms to account for strict type checking which doesn't allow implicit BigInt <-> number coercion.
This also attempts to take advantage of TS specific things like declaration merging.

**Comment by Theaninova** [comment]
> Thanks @Theaninova for the PR! I've an incomplete ground-up implementation from the original work from fudgepop at [this compiler fork](https://github.com/nullableVoidPtr/kaitai_struct_compiler/tree/TypeScriptIntegration) and [runtime](https://github.com/nullableVoidPtr/kaitai_struct_typescript_runtime). I've attempted to shoe in BigInt, but it seems there would need to be a lot of Kaitai Ast transforms to account for strict type checking which doesn't allow implicit BigInt <-> number coercion. This also attempts to take advantage of TS specific things like declaration merging, but all of it is wholly untested

Hmm, I've got a pretty much complete version of the compiler & runtime that I've been using in [this project](https://github.com/Theaninova/mmtorg/tree/master/src/lib/forge/kaitai) for a few days. I wasn't aware that there was already another attempt, though I've been working around some of the type safety issues by wrapping a few things in `( as any)`

**Comment by nullableVoidPtr** [comment]
It's far more involved and I actually had this on the backburner for a little under a year now. Besides the `BigInt` stuff there's no type errors and it's strictly typed according to `deno check`. Of course, runtime verification is another thing :)

**Comment by Theaninova** [comment]
Oh interesting, I'll give it a run

**Comment by fudgepop01** [comment]
(aside: all this discussion makes it clear that I really need to get a handle on my github notifications ahahah 😅 )

It's really cool that @Theaninova made a version that compiles against all tests!! If desired, I'll see if I can plug that into the vscode i'm presently updating (now with a better hex editor that actually works well with scroll wheels...) - The work you've done @nullableVoidPtr is awesome too! The base I made, while it technically worked for some files, sadly did not cover all cases.

Here's @Theaninova's PR on the compiler repo: kaitai-io/kaitai_struct_compiler#249

**Comment by Azq2** [comment]
I spent a couple of days digging into kaitai and came to the following conclusions.

Overall, this is how I see proper and ideal TypeScript support:

* A new compiler for TypeScript (generate .ts directly, do not even think about .d.ts)
* A new compiler for JavaScript (inherits from the TypeScript compiler, but strips types and provides fallback for enum/namespace)
* Serialization support should be implemented (otherwise what is the point?)

Key points:

* No ES5, UMD, or CommonJS. Only ES6, ESM, and TypeScript.
* Named exports
* NO `any` or `unknown`
* The kaitai-struct module should be rewritten to use ESM modules (including "type: module" in package.json) instead of CommonJS or UMD
* The kaitai-struct module should be extended to support serialization
* Classes should inherit from a common base class, as in other languages
* Consistent null/undefined handling: always use undefined (more appropriate for TypeScript)
* The old JavaScript compiler should either be removed or renamed to javascript_es5 if it is still needed for kaitai web

What is required (minimum):

* Implement new TS/JS compilers in kaitai_struct_compiler (use Python as a reference)
* Implement serialization support in kaitai-struct module + compilation to ESM by TSC
* Implement a translator and infrastructure for TS/JS in kaitai_struct_tests (use Python as a reference) <-- MOST IMPORTANT, TO BE SURE CORRECT WORK
* Wait for `+Infinity` years until the PR is accepted :) (≈ never)

**Comment by Azq2** [comment]
Some interesting implementation details... Since we want RW support, we now have to use getters/setters.
The only reason for this - after changing any member we need to mark the model as dirty.

```TypeScript
/**
* Branch instruction jumping into the boot code
*/
// Getters/Setters
get jumpInstruction() {
	return this.#jumpInstruction;
}

set jumpInstruction(v: number) {
	this._setDirty(true); // protected method from KaitaiStruct
	this.#jumpInstruction = v;
}
...

// Attributes declarations
#jumpInstruction!: number; // private data don't seen outside object
...
```

Also, this approach has a nice side effect - the structure is now truly read-only in RO mode. Just like in Java/PHP and other languages (but that's actually a good thing).

---
If we dream a bit - it would be great to have a fully functional interface (as an option) for parsing and serialization.
This approach would be much better for TypeScript:
- Interfaces that can be used directly in code
- Plain objects as input/output - clean data, maximum transparency
- No need to use getters/setters
- Stateless!!!!

OOP only adds overhead (including big mental overhead).

```TypeScript
export interface EgonPublicBootHeader {
	jumpInstruction!: number;
	...
}
export function parseEgonPublicBootHeader(input: Buffer | KaitaiStream): EgonPublicBootHeader {
	...
}
export function serializeEgonPublicBootHeader(data: EgonPublicBootHeader): Buffer {
	...
}
export function checkEgonPublicBootHeader(data: EgonPublicBootHeader): void {
	...
}
```

**Comment by Azq2** [comment]
Also, regarding the processors - the current implementation is inconsistent and hard to understand.
After some investigation, I think it should work as follows:

**OLD LOGIC (CJS/UMD):**
`some.module.my_custom_fx(0x20, false, [0])` -> `require("MyCustomFx").MyCustomFx` // Why not local file????
`some_module.my_custom_fx(0x20, false, [0])` -> `require("some_module/MyCustomFx").MyCustomFx` // from external module
`some.module.my_custom_fx(0x20, false, [0])` -> `require("some-module/MyCustomFx").MyCustomFx` // from external module

Totally incompatible with modern JS/TS.

**NEW LOGIC (ESM):**
`some.module.my_custom_fx(0x20, false, [0])` -> `import { MyCustomFx } from "./MyCustomFx.js"` // local file, `.js` extension is TS/ESM compatible
`some_module.my_custom_fx(0x20, false, [0])` -> `import { MyCustomFx } from "some_module"` // from external module
`some.module.my_custom_fx(0x20, false, [0])` -> `import { MyCustomFx } from "some-module"` // from external module

Compatible with modern JS/TS. Two ways to include a processor - from a module or from a local file. Logic is consistent with Java/Python.

Sorry for the monologue. I hope some Kaitai maintainers will read this and correct me before I complete the TS integration. :)

**Comment by Azq2** [comment]
I have some progress with a fully-featured TypeScript compiler.

Current code (if anyone is interested): https://github.com/kaitai-io/kaitai_struct_compiler/compare/master...yet-another-org-with-forks:kaitai_struct_compiler_typescript:typescript-support

This can generate syntactically correct TypeScript code. Example with Gif.ksy: https://gist.github.com/Azq2/01611c88b2637f1e5667062db2c42110

But:
- At this moment it is just a PoC/Prototype.
- This is only the compiler. The runtime is not currently implemented.
- It has not been tested on all available .ksy files in the formats/tests repo.

Some important implementation notes:
- We should use the null-assertion operator (`!`) for all member accesses inside `_write`/`_read`/`_check`.
- All private attributes are declared as private members (`#`). This is a nice feature, because they are not visible outside the object. The logic for private/public is synchronized with Java.
- Nullable members are declared like `#_root?: Gif;`, otherwise `#magic!: Uint8Array;`
- All access to attributes is implemented using getters/setters.
- Type downcasting as in Java is mandatory. Type casting logic is synchronized with Java.
- Generation code for serialization support has been implemented.

Notes for future work:
- kaitai-struct should be backward-compatible with the legacy CJS/UMD variant.
- `bigint` must be implemented and enabled by default for TypeScript.
  - This requires additional methods for `bigint` in kaitai-struct (due to backward compatibility with legacy).
  - Maybe an optional flag `--typescript-bigint true/false` would be useful.
- Old JavaScript compiler is still needed (e.g. for the web IDE) and should not be removed.
  I think it is better to add ONLY a TypeScript compiler, but with an additional flag `--typescript-strip-types true/false`.
  With this approach we can achieve modern TypeScript/JavaScript with a single fully-featured compiler without breaking legacy projects.
  UPD: I found comment https://github.com/kaitai-io/kaitai_struct/issues/539#issuecomment-473729127
  Should added as typescript + javascript_es6, without removing old javascript


@generalmimon I'm on right way?

**Comment by generalmimon** [comment]
@Azq2:

> Current code (if anyone is interested): https://github.com/kaitai-io/kaitai_struct_compiler/compare/master...yet-another-org-with-forks:kaitai_struct_compiler_typescript:typescript-support

At first glance, it doesn't look bad. It's refreshing to see someone trying to keep things consistent with other target languages instead of reinventing the wheel (which almost all external contributors tend to do).

> * A new compiler for TypeScript (generate .ts directly, do not even think about .d.ts)

Makes sense.

> * A new compiler for JavaScript (inherits from the TypeScript compiler, but strips types and provides fallback for enum/namespace)

In any case, it makes no sense to maintain two separate implementations (`TypeScriptCompiler` and `JavaScriptCompiler`), which would be almost identical except for the typing. I'm not sure whether it's more appropriate for `JavaScriptCompiler` to inherit from `TypeScriptCompiler`, or the other way around.

Maybe `TypeScriptCompiler` should inherit from `JavaScriptCompiler`, since it mostly adds on top of it. If it were the other way around, then `JavaScriptCompiler` would have to _remove_ things, which seems harder and potentially more error-prone (you don't want to forget to override some TypeScript-specific implementation and then end up with invalid JavaScript code when using `-t javascript`).

> * Old JavaScript compiler is still needed (e.g. for the web IDE) and should not be removed.
>   I think it is better to add ONLY a TypeScript compiler, but with an additional flag `--typescript-strip-types true/false`.
>   With this approach we can achieve modern TypeScript/JavaScript with a single fully-featured compiler without breaking legacy projects.

I'm not entirely sure what you mean, but if you propose generating "modern JavaScript" with `-t typescript` + `--typescript-strip-types true`, then I disagree: `-t javascript` (generating `.js` files) vs. `-t typescript` (generating `.ts` files) is much more intuitive.

> * kaitai-struct should be backward-compatible with the legacy CJS/UMD variant.

Perhaps it doesn't have to be, it's probably safe to switch to ESM at this point. AFAIK ESM are part of ES2015 (ES6), which is nowadays widely supported and it feels like it's already deemed "too old" by much of the JS community. Also, TypeScript 6 [deprecates support for the ES5 target](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-6-0.html#deprecated-target-es5) that we're currently [using in the runtime library](https://github.com/kaitai-io/kaitai_struct_javascript_runtime/blob/16de5b1b5b3a8b44a2d3d49b14422df1f9c03995/tsconfig.json#L14), so it makes sense to change the target to at least ES2015 (ES6) for this reason too.

> * `bigint` must be implemented and enabled by default for TypeScript.
>   
>   * This requires additional methods for `bigint` in kaitai-struct (due to backward compatibility with legacy).
>   * Maybe an optional flag `--typescript-bigint true/false` would be useful.

This has nothing to do with TypeScript, it's a JavaScript issue in general: https://github.com/kaitai-io/kaitai_struct/issues/183

I will actually deal with this in the near future, so I recommend ignoring this problem to avoid unnecessary work. As I laid out in https://github.com/kaitai-io/kaitai_struct_webide/issues/189#issuecomment-3101879034, I plan to switch from `number` to `BigInt` only for integers that actually need it (i.e. more than 53 bits of precision or bitwise operations beyond 32 bits are required) - so most integers will stay as `number`. I don't plan to support anything other than the native `BigInt` type - now that it's well supported, third-party libraries that emulate it are fortunately a thing of the past.

**Comment by Azq2** [comment]
@generalmimon Thanks for important notes.

At this moment, `JavaScriptCompiler extends TypeScriptCompiler` is simpler than `TypeScriptCompiler extends JavaScriptCompiler`.

That's because:
- TypeScript is designed for stripping types (https://devblogs.microsoft.com/typescript/announcing-typescript-5-8-beta/#the---erasablesyntaxonly-option)
- All type-related syntax will be moved to a few `renderXXX` functions which can be simply overridden in `JavaScriptCompiler`
- I think TS and JS should be separately tested in CI as different languages, so we cannot get broken code

Current real JS vs TS differences:
- No type declarations
- No not-null in member access
- Namespaces replaced with nested classes (`static InnerName = class extends KaitaiStruct {`)
- Enums replaced with `Object.freeze` (`static EnumName = Object.freeze({`)

Examples of implementation:
- https://github.com/yet-another-org-with-forks/kaitai_struct_compiler_typescript/blob/typescript-support/shared/src/main/scala/io/kaitai/struct/languages/JavaScriptCompiler.scala
- https://github.com/yet-another-org-with-forks/kaitai_struct_compiler_typescript/blob/typescript-support/shared/src/main/scala/io/kaitai/struct/translators/JavaScriptTranslator.scala

----

**Possible alternative variant:** both `TypeScriptCompiler` and `JavaScriptCompiler` extends `ECMAScriptCompiler`,
where `ECMAScriptCompiler` implements all the complex logic (including types) and `TypeScriptCompiler`/`JavaScriptCompiler` contains only a few methods for handling the different syntax.

**Comment by Azq2** [comment]
Some investigations regarding runtime...

###  JavaScript Standard

Some great features worth adopting - private class fields:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_elements
- https://caniuse.com/?search=private+fields

**Pros:**
- Private fields don't pollute the object, making debugging with `console.log` easier.
- This is a baseline feature (95.58%), available in all modern browsers.
- Supported by Node.js 12 (EOL in 2027), which is currently the minimum target for kaitai-struct.

**Cons:**
- We would need to target ES2022. Despite all browsers and engines from 2019–2021 already supporting private fields, the feature was only formally added in ES2022.

My opinion - ES2022 is a good choice:
- The dark era of IE is over. All browsers are regularly auto-updated, so supporting 4+ year old standards no longer makes sense.
- Anyone who needs legacy support can still compile to any older standard (e.g. with Babel) - even ES5.
- Browser support is excellent

---

### What can be removed

`TextDecoder` is now a baseline feature (96.72%):
- https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder
- https://caniuse.com/?search=TextDecoder

The `iconv-lite`/`Buffer` polyfill in `bytesToStr`/`strToBytes` is no longer needed.

---

### ESM pitfalls

Some current implementations don't fit well into the ESM model:
- Synchronous `require()` is no longer available.
- `import()` is always async (whereas `KaitaiStream` is designed for sync use).
- True ESM cannot rely on globals.

This makes lazy auto-loading of `zlib` no longer possible:

```typescript
/**
 * @param buf The input bytes.
 * @returns The uncompressed bytes.
 */
public static processZlib(buf: Uint8Array): Uint8Array {
  if (typeof require !== 'undefined') {
    // require is available - we're running under node
    if (typeof KaitaiStream.zlib === 'undefined')
      KaitaiStream.zlib = require('zlib') as Zlib; // <-- incompatible in ESM
    // use node's zlib module API
    const r = (KaitaiStream.zlib as Zlib).inflateSync(
        Buffer.from(buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength))
    );
    return new Uint8Array(r.buffer, r.byteOffset, r.length);
  } else {
    // no require() - assume we're running as a web worker in browser.
    // user should have configured KaitaiStream.depUrls.zlib, if not
    // we'll throw.
    if (typeof KaitaiStream.zlib === 'undefined'
      && typeof KaitaiStream.depUrls.zlib !== 'undefined') {
      importScripts(KaitaiStream.depUrls.zlib);
      KaitaiStream.zlib = pako; // <-- incompatible in ESM
    }
    // use pako API
    return (KaitaiStream.zlib as Pako).inflate(buf);
  }
}
```

A better approach is to replace this with a generic zlib handler instead of hardcoding `pako`/`zlib`:

```typescript
namespace KaitaiStream {
  interface ZlibHandler {
    inflate(buf: Uint8Array): Uint8Array;
    deflate(buf: Uint8Array): Uint8Array;
  }
}

class KaitaiStream {
  static zlibHandler?: KaitaiStream.ZlibHandler;
  // ....
}
```

The library user can then set any zlib implementation they prefer:

```typescript
import { KaitaiStream } from 'kaitai-struct';
import zlib from "node:zlib";

KaitaiStream.zlibHandler = {
  inflate(buf: Uint8Array): Uint8Array {
    const r = zlib.inflateSync(
      Buffer.from(buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength))
    );
    return new Uint8Array(r.buffer, r.byteOffset, r.length);
  },
  deflate(buf: Uint8Array): Uint8Array {
    const r = zlib.deflateSync(
      Buffer.from(buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength))
    );
    return new Uint8Array(r.buffer, r.byteOffset, r.length);
  }
};
```

It's also still possible to auto-load a `node:zlib` implementation for Node.js environments via `package.json` entrypoints:

**package.json**
```json
{
  "main": "dist/index.js",        // sets zlibHandler using node:zlib
  "browser": "dist/index.browser.js", // does not set zlibHandler
  "types": "dist/index.d.ts"
}
```

https://docs.npmjs.com/cli/v11/configuring-npm/package-json#browser

This field is respected by all major bundlers and by esm.sh.

**Comment by Azq2** [comment]
And now.... I'm frustrated. JavaScript is painful :(
I found a really big problem which affects developer experience.

If we use getters for attribute access:
- Getters are non-enumerable for classes. So console.log shows an empty object:
  ```
  console.log(gif):
  Gif {
    _io: ...,
  }

  console.log(Object.keys(gif)):
  ["_io"]
  ```
- We can patch the object proto to make it enumerable=true, but then we see something like this:
  ```
  console.log(gif):
  Gif {
    _io: ...,
    hdr: [Getter],
    logicalScreenDescriptor: [Getter],
    globalColorTable: [Getter],
    blocks: [Getter],
    _raw_globalColorTable: [Getter]
  }

  console.log(Object.keys(gif)):
  ["_io", "hdr", "logicalScreenDescriptor", "globalColorTable", "blocks", "_raw_globalColorTable"]
  ```

	Still bad, because we don't see the real values, only the strange `[Getter]` placeholder.

With getters, we can't properly debug KaitaiStruct objects during development, and I don't know of a good way to fix this.

A compromise is to use a Proxy - a similar approach is already used in the Python runtime: https://github.com/kaitai-io/kaitai_struct_python_runtime/blob/626d6967159858a98f6f44d2df81d97c44249eac/kaitaistruct.py#L82

1. We can do something like this in base KaitaiStruct class:
    ```typescript
    protected constructor(io?: KaitaiStream) {
      super(io);
    
      return new Proxy(this, {
        set: (target, prop, value, receiver) => {
          if (typeof prop === "string")) {
            if ( // this can be enhanced with generated _writableFields = new Set([...])
              !prop.startsWith("_") ||
              prop.startsWith("_unnamed") ||
              prop.startsWith("_raw") ||
              prop === "_root" || prop === "_parent"
            ) {
              Reflect.set(target, "_dirty", true);
            }
          }
          return Reflect.set(target, prop, value, receiver);
        }
      });
    }
    ```
2. All attributes should declared as direct class members:
    ```typescript
    #_m_transparentColorFlag?: boolean | undefined;
    #_m_userInputFlag?: boolean | undefined;
    blockSize!: Uint8Array;
    flags!: number;
    delayTime!: number;
    transparentIdx!: number;
    terminator!: Uint8Array;
    _root?: Gif | undefined;
    _parent?: Gif.Extension | undefined;
    ```
3. Now we have something like this:
    ```
    console.log(gif):
    <ref *1> Gif {
      hdr: Header {
        magic: Uint8Array(3) [ 71, 73, 70 ],
        version: '89a',
        _root: [Circular *1],
        _parent: [Circular *1]
      },
      ....
    }
    Object.keys(gif):
    [
      'hdr',
      'logicalScreenDescriptor',
      'globalColorTable',
      'blocks',
      '_root',
      '_parent',
      '_raw_globalColorTable'
    ]
    ```

**Comment by Azq2** [comment]
But anyway, I will continue development. All controversial issues can be resolved after sending the first version of the PR.

Some notes about CI/tests.

1. Surprisingly, mocha already supports ESM/TSX. And it works.
    In the future, I think we can migrate to the native NodeJS test infrastructure: [https://nodejs.org/api/test.html#test-reporters](https://nodejs.org/api/test.html#test-reporters)
    This feature is available on v18.15.0+, but the minimal NodeJS version is 12.
  
2. The era of `NODE_PATH` is over. This is bad and undefined behavior.
    All NodeJS applications must contain their own `package.json` / `node_modules` at the top level.
    This means that any out-of-tree sources must either:
    - Have their own `package.json` (as a dedicated package)
    - Be moved inside the main package
  
    This is a big problem for the current architecture where we have two independent source trees:
    ```
    spec/javascript/
      package.json
      testHelper.js
      spec/
        TestFormatName.js
        ....
    compiled/javascript/
      FormatName.js
      ...
    ```
  
    At the moment, I use a workaround by symlinking compiled sources inside `spec/javascript/`:
    ```bash
    ln -sf ../../compiled/javascript spec/javascript/compiled
    ```
  
    This works, but only with `NODE_PRESERVE_SYMLINKS=1` ([docs](https://nodejs.org/api/cli.html#node-preserve-symlinks1)).
    Better than nothing.
  
    Alternative approaches:
    - Refactor the entire test system so that all lang-specific things are contained in a single top-level folder
    - Add `package.json` to `compiled/javascript` and import it as a package
  
    I think symlinks are the better option at the moment.
  
3. We have a problem with importing processors. Parsers in `compiled/javascript` have imports like this:
    ```javascript
    import { MyCustomFx } from "./MyCustomFx.js"; // problem
    import { CustomFx } from "nested-deeply";
    ```
    
    This does not work because we cannot place `MyCustomFx.js` into `compiled/javascript/MyCustomFx.js`.
    To resolve this, we should add new options to the compiler - `--javascript-opaque` / `--typescript-opaque`.
    These options should work like `--nim-opaque` in NimCompiler - just adding a top-level path for imports.
  
    Example with `--javascript-opaque "../opaque/"`:
    ```javascript
    import { MyCustomFx } from "../opaque/MyCustomFx.js";
    ```
    
    Example with `--javascript-opaque "#processors/"` ([subpath imports](https://nodejs.org/api/packages.html#subpath-imports)):
    ```javascript
    import { MyCustomFx } from "#processors/MyCustomFx.js";
    ```

**Comment by Azq2** [comment]
`TextEncoder` only supports UTF-8 :)))) <s>I love JavaScript</s>

Possible alternative implementation:
`KaitaiStream.textEncodingHandler` like `KaitaiStream.zlibHandler`

```ts
import iconv from 'iconv-lite';

// User-defined handler
KaitaiStream.textEncodingHandler = {
	encode(input: string, encoding: string): Uint8Array { // called as fallback
		const r = iconv.encode(input, encoding);
		return new Uint8Array(r.buffer, r.byteOffset, r.length);
	},
	decode(input: Uint8Array, encoding: string): string { // called as fallback
		return iconv.decode(input, encoding);
	}
};

// Or KaitaiStream.setTextEncodingHandler
```

For NodeJS, `iconv-lite` can be included by default (library weight is not problem).

**Comment by Azq2** [comment]
Or something like this....

// This is not an out-of-scope problem, because the previous text encoding mechanism relied on CJS `require()` hacks, which are not possible in ESM.

```TypeScript
// index.node.ts
import iconv from 'iconv-lite'; // it's NodeJS, safe to use directly

/**
 * For NodeJS environments, we use the external iconv-lite module.
 */
KaitaiStream.addTextEncodingHandler({
  encode(str, encoding) {
    if (!iconv.encodingExists(encoding))
      return undefined;
    return iconv.encode(str, encoding);
  },
  decode(buf, encoding) {
    if (!iconv.encodingExists(encoding))
      return undefined;
    return iconv.decode(buf, encoding);
  }
});

```

```TypeScript

  public static textEncodingHandlers: KaitaiStream.TextEncodingHandler[] = [
    // ASCII
    {
      encode(str, encoding) {
        if (encoding !== 'ascii')
          return undefined;
        return KaitaiStream.createBytesFromString(str);
      },
      decode(bytes, encoding) {
        if (encoding !== 'ascii')
          return undefined;
        return KaitaiStream.createStringFromArray(bytes);
      }
    },
    // TextEncoder & TextDecoder
    {
      encode(str, encoding) {
        if (typeof TextEncoder !== 'function')
          return undefined;
        if (!['utf-8', 'utf8'].includes(encoding))
          return undefined;
        return new TextEncoder().encode(str);
      },
      decode(bytes, encoding) {
        if (typeof TextDecoder !== 'function')
          return undefined;
        let decoder: TextDecoder;
        try {
          decoder = new TextDecoder(encoding);
        } catch {
          // encoding is not supported
          return undefined;
        }
        return decoder.decode(bytes);
      }
    },
    // NodeJS Buffer
    {
      encode(str, encoding) {
        if (typeof Buffer !== 'function' || typeof Buffer.isEncoding !== 'function')
          return undefined;
        if (!Buffer.isEncoding(encoding))
          return undefined;
        const r = Buffer.from(str, encoding);
        return new Uint8Array(r.buffer, r.byteOffset, r.byteLength);
      },
      decode(bytes, encoding) {
        if (typeof Buffer !== 'function' || typeof Buffer.isEncoding !== 'function')
          return undefined;
        if (!Buffer.isEncoding(encoding))
          return undefined;
        return Buffer.from(bytes).toString(encoding);
      }
    }
  ];

```
